File: SpectrumAnalyzer/spectrum/Spectrum.cs
using System;
using System.Runtime.InteropServices;
using System.Threading.Tasks;

namespace Spectrum {
	public partial class Spectrum : IDisposable {
		/// <summary>サンプリング周波数[Hz]</summary>
		readonly int SampleRate;

		/// <summary>基本周波数(C0)[Hz]</summary>
		public static double BaseFreq;

		/// <summary>トランスポーズ[半音]</summary>
		public double Transpose { get; set; } = 0.0;

		/// <summary> 最大値 </summary>
		public double Max { get; private set; } = 1.0;

		/// <summary> 自動ゲイン </summary>
		public double AutoGain { get; private set; } = 1.0;

		/// <summary>表示用ピーク</summary>
		public double[] Peak { get; private set; }

		/// <summary>表示用曲線</summary>
		public double[] Curve { get; private set; }

		/// <summary>波形合成用ピーク</summary>
		internal PeakBank[] PeakBanks { get; private set; }

		/// <summary>帯域通過フィルタバンク</summary>
		IntPtr[] mpFilterBanks;

		public Spectrum(int sampleRate) {
			SampleRate = sampleRate;
			Peak = new double[BANK_COUNT];
			Curve = new double[BANK_COUNT];
			PeakBanks = new PeakBank[BANK_COUNT];
			mpFilterBanks = new IntPtr[BANK_COUNT];
			BaseFreq = 442 * Math.Pow(2, HALFTONE_CENTER / OCT_DIV + 3 / 12.0 - 5);
			for (int b = 0; b < BANK_COUNT; ++b) {
				var frequency = BaseFreq * Math.Pow(2.0, (b - 0.5 * HALFTONE_DIV) / OCT_DIV);
				PeakBanks[b] = new PeakBank() {
					DELTA = frequency / SampleRate
				};
				mpFilterBanks[b] = Marshal.AllocHGlobal(Marshal.SizeOf<FilterBank>());
				SetBPF(b, frequency);
			}
		}

		public void Dispose() {
			foreach (var pFilterBank in mpFilterBanks) {
				Marshal.FreeHGlobal(pFilterBank);
			}
		}

		/// <summary>
		/// スペクトルを更新
		/// </summary>
		/// <param name="pInput">入力バッファ(float型ポインタ 2ch×サンプル数)</param>
		/// <param name="sampleCount">入力バッファのサンプル数</param>
		public unsafe void Update(IntPtr pInput, int sampleCount) {
			/* パワースペクトルを算出 */
			CalcPower((float*)pInput, sampleCount);
			/* ピークを抽出 */
			ExtractPeak();
		}

		static double GetAlpha(double sampleRate, double frequency) {
			var halfToneWidth = 1.0 + Math.Log(FREQ_AT_HALFTONE_WIDTH / frequency, 2.0);
			if (halfToneWidth < 1.0) {
				halfToneWidth = 1.0;
			}
			var omega = 2 * Math.PI * frequency / sampleRate;
			var s = Math.Sin(omega);
			var x = Math.Log(2) / 4 * halfToneWidth / 12.0 * omega / s;
			var a = s * Math.Sinh(x);
			return a;
		}

		unsafe void SetBPF(int index, double frequency) {
			var omega = 2 * Math.PI * frequency / SampleRate;
			var alpha = GetAlpha(SampleRate, frequency);
			var a0 = 1.0 + alpha;
			var pBank = (FilterBank*)mpFilterBanks[index];
			Marshal.StructureToPtr(new FilterBank(), (IntPtr)pBank, true);
			pBank->KB0 = alpha / a0;
			pBank->KA1 = -2.0 * Math.Cos(omega) / a0;
			pBank->KA2 = (1.0 - alpha) / a0;
			pBank->SIGMA = GetAlpha(SampleRate / Math.PI * 0.5, frequency);
		}

		unsafe void CalcPower(float* pInput, int sampleCount) {
			Parallel.ForEach(mpFilterBanks, ptr => {
				var pWave = pInput;
				var pBank = (FilterBank*)ptr;
				var KB0 = pBank->KB0;
				var KA2 = pBank->KA2;
				var KA1 = pBank->KA1;
				var SIGMA = pBank->SIGMA;
				for (int s = sampleCount; s != 0; --s) {
					/*** 左チャンネル ***/
					{
						/* 帯域通過フィルタに通す */
						var b = *pWave++;
						var a = b - pBank->lb2;
						a *= KB0;
						pBank->lb2 = pBank->lb1;
						pBank->lb1 = b;
						a -= KA2 * pBank->la2;
						a -= KA1 * pBank->la1;
						pBank->la2 = pBank->la1;
						pBank->la1 = a;
						/* パワースペクトルを得る */
						a *= a;
						pBank->l += (a - pBank->l) * SIGMA;
					}
					/*** 右チャンネル ***/
					{
						/* 帯域通過フィルタに通す */
						var b = *pWave++;
						var a = b - pBank->rb2;
						a *= KB0;
						pBank->rb2 = pBank->rb1;
						pBank->rb1 = b;
						a -= KA2 * pBank->ra2;
						a -= KA1 * pBank->ra1;
						pBank->ra2 = pBank->ra1;
						pBank->ra1 = a;
						/* パワースペクトルを得る */
						a *= a;
						pBank->r += (a - pBank->r) * SIGMA;
					}
				}
			});
			/* 最大値を更新 */
			Max = AUTOGAIN_MIN;
			foreach (FilterBank* pBank in mpFilterBanks) {
				var linear = Math.Sqrt(Math.Max(pBank->l, pBank->r) * 2);
				Max = Math.Max(Max, linear);
			}
			/* 最大値に追随して自動ゲインを更新 */
			var autoGainDelta = (double)sampleCount / SampleRate;
			var diff = Max - AutoGain;
			if (diff < 0) {
				autoGainDelta /= AUTOGAIN_DOWN_SPEED;
			} else {
				autoGainDelta /= AUTOGAIN_UP_SPEED;
			}
			AutoGain += diff * autoGainDelta;
		}

		unsafe void ExtractPeak() {
			var lastL = 0.0;
			var lastR = 0.0;
			var lastLIndex = -1;
			var lastRIndex = -1;
			var lastDisp = 0.0;
			var lastDispIndex = -1;
			for (int idxB = 0; idxB < BANK_COUNT; ++idxB) {
				/*** ピーク抽出用の閾値を算出 ***/
				var thresholdL = 0.0;
				var thresholdR = 0.0;
				{
					/* 音域によって閾値幅と閾値ゲインを選択 */
					int width;
					double gain;
					var transposedIdxB = idxB + Transpose * HALFTONE_DIV;
					if (transposedIdxB < END_LOW_BANK) {
						width = THRESHOLD_WIDTH_LOW;
						gain = THRESHOLD_GAIN_LOW;
					} else if (transposedIdxB < BEGIN_MID_BANK) {
						var a2b = (double)(transposedIdxB - END_LOW_BANK) / (BEGIN_MID_BANK - END_LOW_BANK);
						width = (int)(THRESHOLD_WIDTH_MID * a2b + THRESHOLD_WIDTH_LOW * (1 - a2b));
						gain = THRESHOLD_GAIN_MID * a2b + THRESHOLD_GAIN_LOW * (1 - a2b);
					} else {
						width = THRESHOLD_WIDTH_MID;
						gain = THRESHOLD_GAIN_MID;
					}
					/* 閾値幅で指定される範囲の平均値を閾値にする */
					for (int w = -width; w <= width; ++w) {
						var bw = Math.Min(BANK_COUNT - 1, Math.Max(0, idxB + w));
						var b = *(FilterBank*)mpFilterBanks[bw];
						thresholdL += b.l;
						thresholdR += b.r;
					}
					width = width * 2 + 1;
					/* パワー⇒リニア変換した値に閾値ゲインを掛ける */
					var scale = 2.0 / width;
					thresholdL = Math.Sqrt(thresholdL * scale) * gain;
					thresholdR = Math.Sqrt(thresholdR * scale) * gain;
				}
				var bank = *(FilterBank*)mpFilterBanks[idxB];
				/*** 波形合成用のピークを抽出 ***/
				{
					var peak = PeakBanks[idxB];
					peak.L = 0.0;
					peak.R = 0.0;
					var linearL = Math.Sqrt(bank.l * 2);
					var linearR = Math.Sqrt(bank.r * 2);
					if (linearL < thresholdL) {
						if (0 <= lastLIndex) {
							PeakBanks[lastLIndex].L = lastL;
						}
						linearL = 0.0;
						lastL = 0.0;
						lastLIndex = -1;
					}
					if (lastL < linearL) {
						lastL = linearL;
						lastLIndex = idxB;
					}
					if (linearR < thresholdR) {
						if (0 <= lastRIndex) {
							PeakBanks[lastRIndex].R = lastR;
						}
						linearR = 0.0;
						lastR = 0.0;
						lastRIndex = -1;
					}
					if (lastR < linearR) {
						lastR = linearR;
						lastRIndex = idxB;
					}
				}
				/*** 表示用のピークを抽出、曲線を設定 ***/
				{
					var linear = Math.Sqrt(Math.Max(bank.l, bank.r) * 2);
					var threshold = Math.Max(thresholdL, thresholdR);
					if (EnableNormalize) {
						linear /= Max;
						threshold /= Max;
					}
					if (EnableAutoGain) {
						linear /= AutoGain;
						threshold /= AutoGain;
					}
					Curve[idxB] = linear;
					Peak[idxB] = 0.0;
					if (linear < threshold) {
						if (0 <= lastDispIndex) {
							Peak[lastDispIndex] = lastDisp;
						}
						linear = 0.0;
						lastDisp = 0.0;
						lastDispIndex = -1;
					}
					if (lastDisp < linear) {
						lastDisp = linear;
						lastDispIndex = idxB;
					}
				}
			}
			if (0 <= lastLIndex) {
				PeakBanks[lastLIndex].L = lastL;
			}
			if (0 <= lastRIndex) {
				PeakBanks[lastRIndex].R = lastR;
			}
			if (0 <= lastDispIndex) {
				Peak[lastDispIndex] = lastDisp;
			}
		}
	}
}

File: pytest/spectrum.py
import math

class _BPFBank:
    KB0: float
    KA1: float
    KA2: float
    SIGMA: float

    a1: float = 0.0
    a2: float = 0.0
    b1: float = 0.0
    b2: float = 0.0
    power: float = 0.0

    def __init__(self, sampleRate: int, freq: float):
        alpha = self.__GetAlpha(sampleRate, freq)
        omega = 2 * math.pi * freq / sampleRate
        a0 = 1.0 + alpha
        self.KB0 = alpha / a0
        self.KA1 = -2.0 * math.cos(omega) / a0
        self.KA2 = (1.0 - alpha) / a0
        self.SIGMA = self.__GetAlpha(sampleRate / 2, freq)

    def __GetAlpha(self, sampleRate: int, freq: float):
        MIN_WIDTH = 1.0
        MIN_WIDTH_AT_FREQ = 1000
        halfToneWidth = MIN_WIDTH + math.log(MIN_WIDTH_AT_FREQ / freq, 2.0)
        if (halfToneWidth < MIN_WIDTH):
            halfToneWidth = MIN_WIDTH
        omega = 2 * math.pi * freq / sampleRate
        s = math.sin(omega)
        x = math.log(2, math.exp(1)) / 4 * omega / s * halfToneWidth / 12.0
        return s * math.sinh(x)

class Spectrum:
    SAMPLE_RATE: int 
    BUFFER_SAMPLES: int
    BANK_COUNT: int
    OCT_DIV: int
    DB_MIN: float = -40

    __AMP_MIN: float = math.pow(10, DB_MIN/20)
    __BASE_FREQ: float = 13.75
    __LOW_FREQ_MAX: int = 80
    __MID_FREQ_MIN: int = 220
    __LOW_TONE_MAX: int
    __MID_TONE_MIN: int
    __LOW_TONE_GAIN: float = 2.0
    __MID_TONE_GAIN: float = 1.0
    __LOW_TONE_WIDTH: int
    __MID_TONE_WIDTH: int

    __BPFBanks: list[_BPFBank] = []

    InputBuffer: list[float] = []
    Amp: list[float] = []
    Threshold: list[float] = []
    Peak: list[float] = []

    def __init__(self, sampleRate: int, bufferLength: int, bankCount: int, octDiv: int):
        # 設定値
        self.SAMPLE_RATE = sampleRate
        self.BUFFER_SAMPLES = bufferLength
        self.BANK_COUNT = bankCount
        self.OCT_DIV = octDiv
        self.__LOW_TONE_WIDTH = int(octDiv * 5 / 12)
        self.__MID_TONE_WIDTH = int(octDiv * 1 / 12)
        self.__LOW_TONE_MAX = int(octDiv * math.log(self.__LOW_FREQ_MAX / self.__BASE_FREQ, 2))
        self.__MID_TONE_MIN = int(octDiv * math.log(self.__MID_FREQ_MIN / self.__BASE_FREQ, 2))
        # 入力バッファを作成
        for i in range(bufferLength):
            self.InputBuffer.append(0)
        # フィルタバンクの作成
        for i in range(bankCount):
            freq = self.__BASE_FREQ * math.pow(2, i / octDiv)
            self.__BPFBanks.append(_BPFBank(sampleRate, freq))
            self.Amp.append(0)
            self.Threshold.append(0)
            self.Peak.append(0)

    def __calcThreshold(self):
        lastAmp = self.DB_MIN
        lastAmpIndex = -1
        for idxB in range(self.BANK_COUNT):
            # 閾値幅と閾値ゲインを設定
            gain: float
            width: int
            if (idxB < self.__LOW_TONE_MAX):
                gain = self.__LOW_TONE_GAIN
                width = self.__LOW_TONE_WIDTH
            elif (idxB < self.__MID_TONE_MIN):
                a2b = (idxB - self.__LOW_TONE_MAX) / (self.__MID_TONE_MIN - self.__LOW_TONE_MAX)
                gain = self.__MID_TONE_GAIN * a2b + self.__LOW_TONE_GAIN * (1 - a2b)
                width = int(self.__MID_TONE_WIDTH * a2b + self.__LOW_TONE_WIDTH * (1 - a2b))
            else:
                gain = self.__MID_TONE_GAIN
                width = self.__MID_TONE_WIDTH
            # 閾値幅で指定される範囲の平均値を閾値にする
            threshold: float = 0.0
            for divB in range(-width, width):
                bd = min(self.BANK_COUNT - 1, max(0, idxB + divB))
                threshold += self.__BPFBanks[bd].power
            threshold /= width*2 + 1
            # パワー⇒リニア⇒db変換後、閾値ゲインを掛ける
            threshold = math.sqrt(threshold)
            if (threshold < self.__AMP_MIN):
                threshold = self.__AMP_MIN
            threshold = 20*math.log10(threshold) + gain
            self.Threshold[idxB] = threshold
            # ピークを抽出
            self.Peak[idxB] = self.DB_MIN
            amp = self.Amp[idxB]
            if (amp < threshold):
                if (0 <= lastAmpIndex):
                    self.Peak[lastAmpIndex] = lastAmp
                amp = self.DB_MIN
                lastAmp = self.DB_MIN
                lastAmpIndex = -1
            if (lastAmp < amp):
                lastAmp = amp
                lastAmpIndex = idxB
        if (0 <= lastAmpIndex):
            self.Peak[lastAmpIndex] = lastAmp

    def exec(self):
        for idxB in range(self.BANK_COUNT):
            bank = self.__BPFBanks[idxB]
            for idxT in range(self.BUFFER_SAMPLES):
                # 帯域通過フィルタに通す
                b0 = self.InputBuffer[idxT]
                a0 = bank.KB0 * b0 - bank.KB0 * bank.b2 - bank.KA1 * bank.a1 - bank.KA2 * bank.a2
                bank.a2 = bank.a1
                bank.a1 = a0
                bank.b2 = bank.b1
                bank.b1 = b0
                # パワースペクトルを得る
                bank.power += (a0 * a0 - bank.power) * bank.SIGMA
            # パワー⇒リニア⇒db変換
            amp = math.sqrt(bank.power)
            if (amp < self.__AMP_MIN):
                amp = self.__AMP_MIN
            self.Amp[idxB] = 20*math.log10(amp)
        # 閾値を格納
        self.__calcThreshold()
